Add reference for asm-goto (#1693)
rust-lang/reference: src/inline-assembly.md

  - The type of the expression may be any integer type, but defaults to `i32` just like integer literals.
  - The value of the expression is formatted as a string and substituted directly into the asm template string.

r[asm.operand-type.supported-operands.label]
* `label <block>`

>   - The address of the block is substituted into the asm template string. The assembly block may jump to the substituted addresses.

Reviewer: ```suggestion
  - The address of the block is substituted into the asm template string. The assembly block may jump to the substituted address.
```

  - After execution of the block, the `asm!` expression returns.
  - The type of the block must be unit or `!` (never).

>   - The block starts new safety context; despite the outer `unsafe` needed for `asm!`, you need extra unsafe to perform unsafe operations inside the block.

Reviewer 2: ```suggestion
  - The block starts new safety context; despite the outer `unsafe` needed for `asm!`, you need an extra `unsafe` to perform unsafe operations inside the block.
```

>   - The block starts new safety context; despite the outer `unsafe` needed for `asm!`, you need an extra `unsafe` to perform unsafe operations inside the block.

Reviewer: ```suggestion
  - The block starts a new safety context; despite the outer `unsafe` block needed for `asm!`, unsafe operations within the `label` block must be wrapped in an inner `unsafe` block.
```

# }
```

r[asm.operand-type.supported-operands.label]
* `label <block>`
  - The address of the block is substituted into the asm template string. The assembly code may jump to the substituted address.
  - After execution of the block, the `asm!` expression returns.
  - The type of the block must be unit or `!` (never).

>   - The block starts a new safety context; despite the outer `unsafe` block needed for `asm!`, unsafe operations within the `label` block must be wrapped in an inner `unsafe` block.

Reviewer 3: ```suggestion
  - The block starts a new safety context: unsafe operations within the `label` block must be wrapped in an inner `unsafe` block, even though the entire `asm!` statement is already wrapped in `unsafe`.
```

- `noreturn`: The `asm!` block never returns, and its return type is defined as `!` (never).
  Behavior is undefined if execution falls through past the end of the asm code.
  A `noreturn` asm block behaves just like a function which doesn't return; notably, local variables in scope are not dropped before it is invoked.

> - When labels are present, `noreturn` means the `asm!` block never fallthrough; the asm block may jump to one of the specified blocks.

Reviewer 2: ```suggestion
- When labels are present, `noreturn` means the execution in the `asm!` block never falls through; the asm block may only exit by jumping to one of the specified blocks.
```

> - When labels are present, `noreturn` means the execution of the `asm!` block never falls through; the asm block may only exit by jumping to one of the specified blocks.

Reviewer: ```suggestion
- When one or more `label` blocks are present, `noreturn` means the execution of the `asm!` block never falls through; the asm block may only exit by jumping to one of the specified blocks.
```

>   The entire `asm!` block will have unit type in this case, unless all label blocks diverge, in which case the return type is `!`.

Reviewer: ```suggestion
  The entire `asm!` block will have unit type in this case, unless all `label` blocks diverge, in which case the return type is `!`.
```

- It is a compile-time error to specify `pure` on an asm block with no outputs or only discarded outputs (`_`).

r[asm.options.checks.noreturn]

> - It is a compile-time error to specify `noreturn` on an asm block with outputs and without labels.

Reviewer: ```suggestion
- It is a compile-time error to specify `noreturn` on an asm block with outputs.
```

If it's a compile-time error to have a `label` block on an asm block with outputs (as below), then whether it's a compile-time error to specify `noreturn` on an asm block with outputs is invariant to whether there are labels, no?

Author: This change is to make `r[asm.options.checks.label-with-outputs]` be the only thing that blocks the `asm-goto-with-outputs` feature.

Reviewer: Make sense, OK.

r[asm.options.checks.label-with-outputs]

> - It is a compile-time error to specify label on an asm block with outputs.

Reviewer: ```suggestion
- It is a compile-time error to have a `label` block on an asm block with outputs.
```

- On arm64ec, [call checkers with appropriate thunks](https://learn.microsoft.com/en-us/windows/arm/arm64ec-abi#authoring-arm64ec-in-assembly) are mandatory when calling functions.

r[asm.rules.only-on-exit]
- The requirement of restoring the stack pointer and non-output registers to their original value only applies when exiting the assembly code.

>   - This means that assembly code that never return (even if not marked `noreturn`) don't need to preserve these registers.

Reviewer 3: Most of the `asm!` block to assembly code changes need some grammar fixups. For example:
```suggestion
  - This means that assembly code that never returns (even if not marked `noreturn`) doesn't need to preserve these registers.
```
